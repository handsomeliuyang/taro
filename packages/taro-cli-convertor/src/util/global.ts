import { ErrCodeMsg } from './index'

interface Globals {
  logFilePath: string
  errCodeMsgs: ErrCodeMsg[]
  currentParseFile: string
}

export const globals: Globals = {
  logFilePath: '',
  errCodeMsgs: [],
  currentParseFile: ''
}
