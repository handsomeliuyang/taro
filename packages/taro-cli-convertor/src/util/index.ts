import {
  fs,
  printLog,
  processTypeEnum,
  promoteRelativePath,
  REG_SCRIPT,
  REG_TYPESCRIPT,
  resolveScriptPath,
} from '@tarojs/helper'
import * as path from 'path'

import type * as t from '@babel/types'

export function getRootPath (): string {
  return path.resolve(__dirname, '../../')
}

export function getPkgVersion (): string {
  return require(path.join(getRootPath(), 'package.json')).version
}

function getRelativePath (rootPath: string, sourceFilePath: string, oriPath: string) {
  // 处理以/开头的绝对路径，比如 /a/b
  if (path.isAbsolute(oriPath)) {
    if (oriPath.indexOf('/') !== 0) {
      return ''
    }
    const vpath = path.resolve(rootPath, oriPath.substr(1))
    if (!fs.existsSync(vpath)) {
      return ''
    }
    let relativePath = path.relative(path.dirname(sourceFilePath), vpath)
    relativePath = promoteRelativePath(relativePath)
    if (relativePath.indexOf('.') !== 0) {
      return './' + relativePath
    }
    return relativePath
  }
  // 处理非正常路径，比如 a/b
  if (oriPath.indexOf('.') !== 0) {
    const vpath = path.resolve(sourceFilePath, '..', oriPath)
    if (fs.existsSync(vpath)) {
      return './' + oriPath
    }
    const testParts = oriPath.split('/')
    const testindex = path.join(rootPath, `node_modules/${testParts[0]}`) // 判断三方件是否在node_modules中
    if (!fs.existsSync(testindex)) {
      if (
        oriPath.indexOf('/') !== -1 &&
        oriPath.indexOf('@') === -1 &&
        oriPath.lastIndexOf('.js') !== oriPath.length - 3
      ) {
        oriPath = oriPath + '.js' // 不在这里返回    utils/auth -> utils/auth.js
      }
      if (fs.existsSync(oriPath)) {
        oriPath = './' + oriPath
      }
    }
  }
  return oriPath
}

function copyFileToTaro (from: string, to: string, options?: fs.CopyOptionsSync) {
  const filename = path.basename(from)
  if (fs.statSync(from).isFile() && !path.extname(to)) {
    fs.ensureDir(to)
    return fs.copySync(from, path.join(to, filename), options)
  }
  fs.ensureDir(path.dirname(to))
  return fs.copySync(from, to, options)
}

export function analyzeImportUrl (
  rootPath: string,
  sourceFilePath: string,
  scriptFiles: Set<string>,
  source: t.StringLiteral,
  value: string,
  external:string[],
  miniprogramRoot: string,
  convertDir: string,
  isTsProject?: boolean
) {
  const valueAbs:string = path.resolve(sourceFilePath, '..' ,value)
  if (external){
    for (const iExternal of external){
      // value绝对路径包含json内绝对路径的dirname,就符合条件,直接copy文件过去
      if(valueAbs.indexOf(path.dirname(iExternal))=== 0){
        const outputFilePath = valueAbs.replace(isTsProject ? miniprogramRoot : rootPath, convertDir)
        copyFileToTaro(valueAbs, outputFilePath)
        // 直接return,不让此文件加入到scriptFiles里后续被转换
        return
      }
    }
  }  
  const valueExtname = path.extname(value)
  const rpath = getRelativePath(rootPath, sourceFilePath, value)
  if (!rpath) {
    printLog(processTypeEnum.ERROR, '引用文件', `文件 ${sourceFilePath} 中引用 ${value} 不存在！`)
    return
  }
  if (rpath !== value) {
    value = rpath
    source.value = rpath
  }
  if (value.indexOf('.') === 0) {
    if (REG_SCRIPT.test(valueExtname) || REG_TYPESCRIPT.test(valueExtname)) {
      const vpath = path.resolve(sourceFilePath, '..', value)
      let fPath = value
      if (fs.existsSync(vpath)) {
        fPath = vpath
      } else {
        printLog(processTypeEnum.ERROR, '引用文件', `文件 ${sourceFilePath} 中引用 ${value} 不存在！`)
      }
      scriptFiles.add(fPath)
    } else {
      let vpath = resolveScriptPath(path.resolve(sourceFilePath, '..', value))
      if (vpath) {
        if (!fs.existsSync(vpath)) {
          printLog(processTypeEnum.ERROR, '引用文件', `文件 ${sourceFilePath} 中引用 ${value} 不存在！`)
        } else {
          if (fs.lstatSync(vpath).isDirectory()) {
            if (fs.existsSync(path.join(vpath, 'index.js'))) {
              vpath = path.join(vpath, 'index.js')
            } else {
              printLog(processTypeEnum.ERROR, '引用目录', `文件 ${sourceFilePath} 中引用了目录 ${value}！`)
              return
            }
          }
          let relativePath = path.relative(sourceFilePath, vpath)
          const relativePathExtname = path.extname(relativePath)
          scriptFiles.add(vpath)
          relativePath = promoteRelativePath(relativePath)
          if (/\.wxs/.test(relativePathExtname)) {
            relativePath += '.js'
          } else if (!isTsProject) {
            relativePath = relativePath.replace(relativePathExtname, '.js')
          }
          source.value = relativePath
        }
      }
    }
  } else {
    if (value.startsWith('/') || value.startsWith('@tarojs')) {
      return
    }
    scriptFiles.add(value)
  }
}

export const incrementId = () => {
  let n = 0
  return () => (n++).toString()
}