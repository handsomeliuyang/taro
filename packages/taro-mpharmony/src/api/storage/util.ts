import native from '../NativeApi'

interface Status {
  done: boolean
  data: string
  errorMsg: string
}

export function getStorageStatus (method: string, key: string) {
  let count = 0
  const res: Status = { done: false, data: '', errorMsg: `${method} timeout` } // false 可能插入数据超时; true, 数据在20000刻度的时间内插入完成
  while (count < 20000) {
    count++
    if (count % 1000 === 0) {
      const status: Status = native.getExecStatus({ method, key })
      if (status.done || status.errorMsg) { // 插入数据完成，退出循环
        return status
      }
    }
  }
  return res
}


export function displayExecRes (status: Status, method: string) {
  if (!status.done) {
    console.error({ errMsg: status.errorMsg })
  } else {
    // eslint-disable-next-line no-console
    console.log(`${method} execution succeeded` + (status.data ? `res: ${status.data}` : ''))
  }
}

export function getItem (data: string) {
  let item
  try {
    item = JSON.parse(data)
    // item = JSON.parse(localStorage.getItem(key) || '')
  } catch (e) {
  } // eslint-disable-line no-empty

  // 只返回使用 Taro.setStorage API 存储的数据
  if (item && typeof item === 'object' && item.hasOwnProperty('data')) {
    return { result: true, data: item.data }
  } else {
    return { result: true, data: '' }
  }
}

export function handleData (data) {
  const type = typeof data
  let obj = {}

  if (type === 'symbol') {
    obj = { data: '' }
  } else {
    obj = { data }
  }
  return obj
}
