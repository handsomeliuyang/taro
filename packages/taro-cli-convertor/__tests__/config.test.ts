import Convertor from '../src'
import { clearMockFiles, getResMapFile, normalizePath, resFileMap, setMockFiles } from './__mocks__/fs-extra'
import { DEMO_JS_FILE_INFO, root } from './data/fileData'
import { removeBackslashesSerializer } from './util'

expect.addSnapshotSerializer(removeBackslashesSerializer)

const pathSystem = require('path')

jest.mock('fs', () => {
  const originalModule = jest.requireActual('fs')
  return {
    ...originalModule,
    promises: {
      ...originalModule.promises,
      writeFile: jest.fn((path, data) => {
        let resPath = ''
        if (pathSystem.isAbsolute(path)) {
          const resArr = normalizePath(path).split(root)
          if (resArr.length === 0) {
            resPath = normalizePath(path).split(root)[0]
          } else {
            resPath = normalizePath(path).split(root)[1]
          }
        }
        if (Buffer.isBuffer(data)) {
          data = Buffer.from(data).toString('utf8')
        }
        resFileMap.set(root + resPath, data)
      }),
      stat: jest.fn(() => {
        return {
          isDirectory: () => true,
        }
      }),
    },
  }
})

describe('Â∞èÁ®ãÂ∫èËΩ¨Êç¢ÁîüÊàêÈÖçÁΩÆÊñá‰ª∂', () => {
  beforeAll(() => {
    jest.spyOn(Convertor.prototype, 'init').mockImplementation(() => {})
  })

  afterEach(() => {
    // Ê∏ÖÁ©∫Êñá‰ª∂‰ø°ÊÅØ
    clearMockFiles()
    jest.restoreAllMocks()
  })

  test('ÈÖçÁΩÆÊñá‰ª∂ÁîüÊàê', (done) => {
    setMockFiles(root, DEMO_JS_FILE_INFO)
    const convertor = new Convertor(root, false)
    convertor.framework = 'react'
    convertor.generateConfigFiles()
    setTimeout(() => {
      expect(resFileMap).toMatchSnapshot()
      done()
    }, 4500)
  })
})

describe('Êó•Âøó', () => {
  beforeAll(() => {
    // mockÊä•ÂëäÁîüÊàê
    jest.spyOn(Convertor.prototype, 'generateReport').mockImplementation(() => {})

    // ÈÖçÁΩÆÊñá‰ª∂ÁîüÊàê
    jest.spyOn(Convertor.prototype, 'generateConfigFiles').mockImplementation(() => {})
  })

  afterEach(() => {
    // Ê∏ÖÁ©∫Êñá‰ª∂‰ø°ÊÅØ
    clearMockFiles()
  })

  afterAll(() => {
    jest.restoreAllMocks()
  })

  test('ÁîüÊàêÊó•Âøó', () => {
    setMockFiles(root, DEMO_JS_FILE_INFO)
    const convertor = new Convertor(root, false)
    convertor.run()
    const resFileMap = getResMapFile()
    expect(resFileMap.get('/wxProject/taroConvert/.convert/convert.log')).toMatchSnapshot()
  })
})

describe('ËΩ¨Êç¢Êä•Âëä', () => {
  beforeAll(() => {
    // ÈÖçÁΩÆÊñá‰ª∂ÁîüÊàê
    jest.spyOn(Convertor.prototype, 'generateConfigFiles').mockImplementation(() => {})
  })

  afterEach(() => {
    // Ê∏ÖÁ©∫Êñá‰ª∂‰ø°ÊÅØ
    clearMockFiles()
  })

  afterAll(() => {
    jest.restoreAllMocks()
  })

  test('ÁîüÊàêËΩ¨Êç¢Êä•ÂëäÊñá‰ª∂', () => {
    const REPORT_DEMO = {
      '/pages': {
        '/index': {
          '/index.js': `
            const app = getApp()
            Page({
              data: {
                motto: 'Hello World',
              },
              onLoad() {}
            })
          `,
          '/index.json': `
            {
              "usingComponents": {}
            }
          `,
          '/index.wxml': `
            <view>
              <text>{{motto}}</text>
            </view>
          `,
          '/index.wxss': '',
        },
      },
      '/project.config.json': `{}`,
      '/app.js': `App({})`,
      '/app.json': `
        {
          "pages":[
            "pages/index/index"
          ]
        }
      `,
    }
    setMockFiles(root, REPORT_DEMO)
    const convertor = new Convertor(root, false)
    convertor.run()
    const resFileMap = getResMapFile()

    expect(resFileMap.has('/wxProject/taroConvert/report')).toBeTruthy()
    expect(resFileMap.has('/wxProject/taroConvert/report/static/js')).toBeTruthy()
    expect(resFileMap.has('/wxProject/taroConvert/report/static/css')).toBeTruthy()
    expect(resFileMap.has('/wxProject/taroConvert/report/static/media')).toBeTruthy()
  })

  describe('ËΩ¨Êç¢Êä•ÂëäËÆ∞ÂΩïÈîôËØØ‰ø°ÊÅØ', () => {
    beforeAll(() => {
      // ÈÖçÁΩÆÊñá‰ª∂ÁîüÊàê
      jest.spyOn(Convertor.prototype, 'generateConfigFiles').mockImplementation(() => {})
    })

    afterEach(() => {
      // Ê∏ÖÁ©∫Êñá‰ª∂‰ø°ÊÅØ
      clearMockFiles()
      jest.restoreAllMocks()
    })

    afterAll(() => {
      jest.restoreAllMocks()
    })

    test('ÂõæÁâáË∑ØÂæÑ‰∏çÂ≠òÂú®', () => {
      const REPORT_DEMO = {
        '/pages': {
          '/index': {
            '/index.js': `
              const app = getApp()
              Page({
                data: {
                  motto: 'Hello World',
                },
                onLoad() {}
              })
            `,
            '/index.json': `
              {
                "usingComponents": {}
              }
            `,
            '/index.wxml': `
              <view>
                <text>{{motto}}</text>
              </view>
              <image src="/images/tutu.jpg" mode=""/>
            `,
            '/index.wxss': '',
          },
        },
        '/images': {},
        '/project.config.json': `{}`,
        '/app.js': `App({})`,
        '/app.json': `
          {
            "pages":[
              "pages/index/index"
            ]
          }
        `,
      }
      setMockFiles(root, REPORT_DEMO)
      const convertor = new Convertor(root, false)
      convertor.run()
      const resFileMap = getResMapFile()

      expect(resFileMap.has('/wxProject/taroConvert/report')).toBeTruthy()
      expect(resFileMap.has('/wxProject/taroConvert/report/static/js')).toBeTruthy()
      expect(resFileMap.has('/wxProject/taroConvert/report/static/css')).toBeTruthy()
      expect(resFileMap.has('/wxProject/taroConvert/report/static/media')).toBeTruthy()
    })

    test('app.json‰∏çÂ≠òÂú®', () => {
      const REPORT_DEMO = {
        '/pages': {
          '/index': {
            '/index.js': `
              const app = getApp()
              Page({
                data: {
                  motto: 'Hello World',
                },
                onLoad() {}
              })
            `,
            '/index.json': `
              {
                "usingComponents": {}
              }
            `,
            '/index.wxml': `
              <view>
                <text>{{motto}}</text>
              </view>
              <image src="/images/tutu.jpg" mode=""/>
            `,
            '/index.wxss': '',
          },
        },
        '/images': {},
        '/project.config.json': `{}`,
        '/app.js': `App({})`,
      }
      setMockFiles(root, REPORT_DEMO)
      jest.spyOn(process, 'exit').mockImplementation()
      const spy = jest.spyOn(console, 'log')
      const convertor = new Convertor(root, false)
      expect(spy).toHaveBeenCalledTimes(3)
      expect(spy.mock.calls[1][0]).toMatchInlineSnapshot(`[31mapp.json ËØªÂèñÂ§±Ë¥•ÔºåËØ∑Ê£ÄÊü•ÔºÅ[39m`)
      convertor.run()
    })
  })
})
