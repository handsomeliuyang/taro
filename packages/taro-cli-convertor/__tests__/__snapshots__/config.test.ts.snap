// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`小程序转换生成配置文件 配置文件生成 1`] = `
Map {
  /wxProject/taroConvert/package.json => {
  "name": "taroConvert",
  "version": "1.0.0",
  "private": true,
  "description": "",
  "templateInfo": {
    "name": "default",
    "typescript": false,
    "css": "sass"
  },
  "scripts": {
    "build:weapp": "taro build --type weapp",
    "build:swan": "taro build --type swan",
    "build:alipay": "taro build --type alipay",
    "build:tt": "taro build --type tt",
    "build:h5": "taro build --type h5",
    "build:rn": "taro build --type rn",
    "build:qq": "taro build --type qq",
    "build:jd": "taro build --type jd",
    "build:quickapp": "taro build --type quickapp",
    "build:mpharmony": "taro build --type mpharmony",
    "build:mpandroid": "taro build --type mpandroid",
    "dev:weapp": "npm run build:weapp -- --watch",
    "dev:swan": "npm run build:swan -- --watch",
    "dev:alipay": "npm run build:alipay -- --watch",
    "dev:tt": "npm run build:tt -- --watch",
    "dev:h5": "npm run build:h5 -- --watch",
    "dev:rn": "npm run build:rn -- --watch",
    "dev:qq": "npm run build:qq -- --watch",
    "dev:jd": "npm run build:jd -- --watch",
    "dev:quickapp": "npm run build:quickapp -- --watch",
    "dev:mpharmony": "npm run build:mpharmony -- --watch",
    "dev:mpandroid": "npm run build:mpandroid -- --watch",
    "test": "jest"
  },
  "browserslist": [
    "last 3 versions",
    "Android >= 4.1",
    "ios >= 8"
  ],
  "author": "",
  "dependencies": {
    "@babel/runtime": "^7.21.5",
    "@tarojs/components": "3.6.15",
    "@tarojs/helper": "3.6.15",
    "@tarojs/plugin-platform-weapp": "3.6.15",
    "@tarojs/plugin-platform-alipay": "3.6.15",
    "@tarojs/plugin-platform-tt": "3.6.15",
    "@tarojs/plugin-platform-swan": "3.6.15",
    "@tarojs/plugin-platform-jd": "3.6.15",
    "@tarojs/plugin-platform-qq": "3.6.15",
    "@tarojs/plugin-platform-h5": "3.6.15",
    "@tarojs/plugin-platform-mpharmony": "3.6.15",
    "@tarojs/plugin-platform-mpandroid": "3.6.15",
    "@tarojs/runtime": "3.6.15",
    "@tarojs/shared": "3.6.15",
    "@tarojs/taro": "3.6.15",
    "@tarojs/plugin-framework-react": "3.6.15",
    "@tarojs/react": "3.6.15",
    "react-dom": "^18.0.0",
    "react": "^18.0.0"
  },
  "devDependencies": {
    "@babel/core": "^7.8.0",
    "@tarojs/cli": "3.6.15",
    "@types/webpack-env": "^1.13.6",
    "@tarojs/test-utils-react": "^0.1.1",
    "@types/react": "^18.0.0",
    "webpack": "5.78.0",
    "@tarojs/taro-loader": "3.6.15",
    "@tarojs/webpack5-runner": "3.6.15",
    "babel-preset-taro": "3.6.15",
    "eslint-config-taro": "3.6.15",
    "eslint": "^8.12.0",
    "@pmmmwh/react-refresh-webpack-plugin": "^0.5.5",
    "react-refresh": "^0.11.0",
    "eslint-plugin-react": "^7.8.2",
    "eslint-plugin-import": "^2.12.0",
    "eslint-plugin-react-hooks": "^4.2.0",
    "stylelint": "^14.4.0",
    "postcss": "^8.4.18",
    "ts-node": "^10.9.1",
    "@types/node": "^18.15.11",
    "@types/jest": "^29.3.1",
    "jest": "^29.3.1",
    "jest-environment-jsdom": "^29.5.0"
  }
}
,
}
`;

exports[`日志 生成日志 1`] = `
INFO [taro-cli-convertor] parseAst - 入参 
{"sourceFilePath":"wxProjectapp.js","outputFilePath":"wxProjectapp.js","importStylePath":null,"imports":[]} 
INFO [taro-cli-convertor] parseAst - 解析ImportDeclaration 
import { Block } from "@tarojs/components"; 
INFO [taro-cli-convertor] analyzeImportUrl - 入参 
sourceFilePath: wxProjectapp.js 
value: @tarojs/components 
INFO [taro-cli-convertor] parseAst - 解析ImportDeclaration 
import React from "react"; 
INFO [taro-cli-convertor] analyzeImportUrl - 入参 
sourceFilePath: wxProjectapp.js 
value: react 
INFO [taro-cli-convertor] parseAst - 解析ImportDeclaration 
import Taro from "@tarojs/taro"; 
INFO [taro-cli-convertor] analyzeImportUrl - 入参 
sourceFilePath: wxProjectapp.js 
value: @tarojs/taro 
INFO [taro-cli-convertor] parseAst - 解析CallExpression 
require("@tarojs/with-weapp") 
INFO [taro-cli-convertor] analyzeImportUrl - 入参 
sourceFilePath: wxProjectapp.js 
value: @tarojs/with-weapp 
INFO [taro-cli-convertor] parseAst - 解析CallExpression 
App({}) 
INFO [taro-cli-convertor] parseAst - 解析CallExpression 
cacheOptions.setOptionsToCache({}) 
INFO [taro-cli-convertor] parseAst - 解析MemberExpression 
cacheOptions.setOptionsToCache 
INFO [taro-cli-convertor] parseAst - 解析ClassDeclaration 
@withWeapp(cacheOptions.getOptionsFromCache(), true)
class App extends React.Component {
  render() {
    return this.props.children;
  }
} 
INFO [taro-cli-convertor] parseAst - 解析MemberExpression 
this.props.children 
INFO [taro-cli-convertor] parseAst - 解析MemberExpression 
this.props 
INFO [taro-cli-convertor] parseAst - 解析MemberExpression 
React.Component 
INFO [taro-cli-convertor] parseAst - 解析CallExpression 
withWeapp(cacheOptions.getOptionsFromCache(), true) 
INFO [taro-cli-convertor] parseAst - 解析CallExpression 
cacheOptions.getOptionsFromCache() 
INFO [taro-cli-convertor] parseAst - 解析MemberExpression 
cacheOptions.getOptionsFromCache 
INFO [taro-cli-convertor] parseAst - 解析StringLiteral 
"@tarojs/with-weapp" 
INFO [taro-cli-convertor] parseAst - 解析StringLiteral 
"@tarojs/components" 
INFO [taro-cli-convertor] parseAst - 解析StringLiteral 
"react" 
INFO [taro-cli-convertor] parseAst - 解析StringLiteral 
"@tarojs/taro" 
INFO [taro-cli-convertor] parseAst - 解析StringLiteral 
"@tarojs/with-weapp" 
INFO [taro-cli-convertor] convertToOptional - 解析ObjectProperty 
cacheOptions 
INFO [taro-cli-convertor] convertToOptional - 解析CallExpression 
require("@tarojs/with-weapp") 
INFO [taro-cli-convertor] convertToOptional - 解析ObjectProperty 
default: withWeapp 
INFO [taro-cli-convertor] convertToOptional - 解析CallExpression 
require("@tarojs/with-weapp") 
INFO [taro-cli-convertor] convertToOptional - 解析CallExpression 
cacheOptions.setOptionsToCache({}) 
INFO [taro-cli-convertor] convertToOptional - 解析ClassBody 
{
  render() {
    return this.props.children;
  }
} 
INFO [taro-cli-convertor] convertToOptional - 解析MemberExpression 
{
  render() {
    return this.props.children;
  }
} 
INFO [taro-cli-convertor] convertToOptional - 解析MemberExpression 
{
  render() {
    return this.props.children;
  }
} 
INFO [taro-cli-convertor] convertToOptional - 解析CallExpression 
withWeapp(cacheOptions.getOptionsFromCache(), true) 
INFO [taro-cli-convertor] convertToOptional - 解析CallExpression 
cacheOptions.getOptionsFromCache() 
INFO [taro-cli-convertor] parseAst - 解析VariableDeclaration 
const {
  cacheOptions
} = require("@tarojs/with-weapp"); 
INFO [taro-cli-convertor] parseAst - 解析VariableDeclaration 
const {
  default: withWeapp
} = require("@tarojs/with-weapp"); 
INFO [taro-cli-convertor] traversePages - 开始转换页面 
pages/index/index 
INFO [taro-cli-convertor] parseAst - 入参 
{"sourceFilePath":"wxProjectpagesindexindex.js","outputFilePath":"wxProjectpagesindexindex.js","importStylePath":null,"depComponents":{},"imports":[],"pluginComponents":{}} 
INFO [taro-cli-convertor] parseAst - 解析ImportDeclaration 
import { Block, View, Text } from "@tarojs/components"; 
INFO [taro-cli-convertor] analyzeImportUrl - 入参 
sourceFilePath: wxProjectpagesindexindex.js 
value: @tarojs/components 
INFO [taro-cli-convertor] parseAst - 解析ImportDeclaration 
import React from "react"; 
INFO [taro-cli-convertor] analyzeImportUrl - 入参 
sourceFilePath: wxProjectpagesindexindex.js 
value: react 
INFO [taro-cli-convertor] parseAst - 解析ImportDeclaration 
import Taro from "@tarojs/taro"; 
INFO [taro-cli-convertor] analyzeImportUrl - 入参 
sourceFilePath: wxProjectpagesindexindex.js 
value: @tarojs/taro 
INFO [taro-cli-convertor] parseAst - 解析CallExpression 
require("@tarojs/with-weapp") 
INFO [taro-cli-convertor] analyzeImportUrl - 入参 
sourceFilePath: wxProjectpagesindexindex.js 
value: @tarojs/with-weapp 
INFO [taro-cli-convertor] parseAst - 解析CallExpression 
Taro.getApp() 
INFO [taro-cli-convertor] parseAst - 解析MemberExpression 
Taro.getApp 
INFO [taro-cli-convertor] parseAst - 解析CallExpression 
Page({
  data: {
    motto: 'Hello World'
  },
  onLoad() {}
}) 
INFO [taro-cli-convertor] parseAst - 解析CallExpression 
cacheOptions.setOptionsToCache({
  data: {
    motto: 'Hello World'
  },
  onLoad() {}
}) 
INFO [taro-cli-convertor] parseAst - 解析MemberExpression 
cacheOptions.setOptionsToCache 
INFO [taro-cli-convertor] parseAst - 解析ClassDeclaration 
@withWeapp(cacheOptions.getOptionsFromCache())
class _C extends React.Component {
  render() {
    const {
      motto
    } = this.data;
    return <View><Text>{motto}</Text></View>;
  }
} 
INFO [taro-cli-convertor] parseAst - 解析MemberExpression 
this.data 
INFO [taro-cli-convertor] parseAst - 解析JSXElement 
<View><Text>{motto}</Text></View> 
INFO [taro-cli-convertor] parseAst - 解析JSXElement 
<Text>{motto}</Text> 
INFO [taro-cli-convertor] parseAst - 解析MemberExpression 
React.Component 
INFO [taro-cli-convertor] parseAst - 解析CallExpression 
withWeapp(cacheOptions.getOptionsFromCache()) 
INFO [taro-cli-convertor] parseAst - 解析CallExpression 
cacheOptions.getOptionsFromCache() 
INFO [taro-cli-convertor] parseAst - 解析MemberExpression 
cacheOptions.getOptionsFromCache 
INFO [taro-cli-convertor] parseAst - 解析StringLiteral 
"@tarojs/with-weapp" 
INFO [taro-cli-convertor] parseAst - 解析StringLiteral 
"@tarojs/components" 
INFO [taro-cli-convertor] parseAst - 解析StringLiteral 
"react" 
INFO [taro-cli-convertor] parseAst - 解析StringLiteral 
"@tarojs/taro" 
INFO [taro-cli-convertor] parseAst - 解析StringLiteral 
"@tarojs/with-weapp" 
INFO [taro-cli-convertor] parseAst - 解析StringLiteral 
'Hello World' 
INFO [taro-cli-convertor] convertToOptional - 解析ObjectProperty 
cacheOptions 
INFO [taro-cli-convertor] convertToOptional - 解析CallExpression 
require("@tarojs/with-weapp") 
INFO [taro-cli-convertor] convertToOptional - 解析ObjectProperty 
default: withWeapp 
INFO [taro-cli-convertor] convertToOptional - 解析CallExpression 
require("@tarojs/with-weapp") 
INFO [taro-cli-convertor] convertToOptional - 解析CallExpression 
Taro.getApp() 
INFO [taro-cli-convertor] convertToOptional - 解析CallExpression 
cacheOptions.setOptionsToCache({
  data: {
    motto: 'Hello World'
  },
  onLoad() {}
}) 
INFO [taro-cli-convertor] convertToOptional - 解析ObjectProperty 
data: {
  motto: 'Hello World'
} 
INFO [taro-cli-convertor] convertToOptional - 解析ObjectProperty 
motto: 'Hello World' 
INFO [taro-cli-convertor] convertToOptional - 解析ClassBody 
{
  render() {
    const {
      motto
    } = this.data;
    return <View><Text>{motto}</Text></View>;
  }
} 
INFO [taro-cli-convertor] convertToOptional - 解析MemberExpression 
{
  render() {
    const {
      motto
    } = this.data;
    return <View><Text>{motto}</Text></View>;
  }
} 
INFO [taro-cli-convertor] convertToOptional - 解析ObjectProperty 
motto 
INFO [taro-cli-convertor] convertToOptional - 解析CallExpression 
withWeapp(cacheOptions.getOptionsFromCache()) 
INFO [taro-cli-convertor] convertToOptional - 解析CallExpression 
cacheOptions.getOptionsFromCache() 
INFO [taro-cli-convertor] parseAst - 解析VariableDeclaration 
const {
  cacheOptions
} = require("@tarojs/with-weapp"); 
INFO [taro-cli-convertor] parseAst - 解析VariableDeclaration 
const {
  default: withWeapp
} = require("@tarojs/with-weapp"); 
INFO [taro-cli-convertor] parseAst - 解析VariableDeclaration 
const app = Taro.getApp(); 
INFO [taro-cli-convertor] parseAst - 解析VariableDeclaration 
const {
  motto
} = this.data; 

`;
